tests/PluginTest.php

```diff
@@ -65,23 +69,81 @@ public function testFileSystem(): void
 
     public function testFunctionExists(): void
     {
-        $plugin = self::loadPlugin("alloc.wasm");
+        $plugin = self::loadPlugin("alloc.wasm", []);
 
         $this->assertTrue($plugin->functionExists("run_test"));
         $this->assertFalse($plugin->functionExists("i_dont_exist"));
     }
 
-    public static function loadPlugin(string $name, ?callable $config = null)
+    public function testHostFunctions(): void
+    {
+        $kvstore = [];
+
+        $kvRead = new HostFunction("kv_read", [ExtismValType::I64], [ExtismValType::I64], function (CurrentPlugin $p, string $key) use (&$kvstore) {
+            return $kvstore[$key] ?? "\0\0\0\0";
+        });
+
+        $kvWrite = new HostFunction("kv_write", [ExtismValType::I64, ExtismValType::I64], [], function (string $key, string $value) use (&$kvstore) {
+            $kvstore[$key] = $value;
+        });
+
+        $plugin = self::loadPlugin("count_vowels_kvstore.wasm", [$kvRead, $kvWrite]);
+
+        $response = $plugin->call("count_vowels", "Hello World!");
+        $this->assertEquals('{"count":3,"total":3,"vowels":"aeiouAEIOU"}', $response);
+
+        $response = $plugin->call("count_vowels", "Hello World!");
+        $this->assertEquals('{"count":3,"total":6,"vowels":"aeiouAEIOU"}', $response);
+    }
+
+    public function testHostFunctionManual(): void
+    {
+        $kvstore = [];
+
+        $kvRead = new HostFunction("kv_read", [ExtismValType::I64], [ExtismValType::I64], function (CurrentPlugin $p, int $keyPtr) use (&$kvstore) {
+            $key = $p->read_block($keyPtr);
+            $value = $kvstore[$key] ?? "\0\0\0\0";
+            return $p->write_block($value);
+        });
+
+        $kvWrite = new HostFunction("kv_write", [ExtismValType::I64, ExtismValType::I64], [], function (CurrentPlugin $p, int $keyPtr, int $valuePtr) use (&$kvstore) {
```

Comment:
(related to the previous comment) 

I suppose the docs I'm looking for are primarily about how the transformation of the "higher level" types into these works. And, what are the limits? If my `string` for example, is actually bytes, is there any issue?

Reply:
AFAIK there is no `byte[]` type in php, and `string` is used in its place in the standard library